File: Blind_75_LeetCode_Questions/c++/String/Longest_Repeating_Character_Replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) 
    {
        unordered_map<char, int> charMap;
        int len = s.length();
        int i = 0;
        int start = 0;
        int maxLength = 0;
        int maxCnt = 0;

        while(i < len)
        {
            charMap[s[i]]++;
            maxCnt = max(maxCnt, charMap[s[i]]);

            if((i - start + 1) - maxCnt > k)
            {
                charMap[s[start]]--;
                start++;
            }
            maxLength = max(maxLength, i - start + 1);
            i++;
        }
        return maxLength;
        
    }
};

File: Blind_75_LeetCode_Questions/java/String/Q424_Longest_Repeating_Character_Replacement/Longest_Repeating_Character_Replacement.java
class Solution {
    public int characterReplacement(String s, int k) 
    {
        int[] letterCnt = new int[26];
        int start = 0;
        int end = 0;
        int maxCnt = 0;
        int len = s.length();

        while(end < len)
        {
            char currentChar = s.charAt(end);
            letterCnt[currentChar - 'A']++;
            maxCnt = Math.max(maxCnt, letterCnt[currentChar - 'A']);
            
            if(end - start + 1 - maxCnt > k)
            {
                letterCnt[s.charAt(start) - 'A']--;
                start++;
            }
            end++;
        }
        return end - start;
    }
}
